[Serialization & ClangImporter] Filter decl to deserialize by their `@objc` attributes

xymus opened the pull request:
Add an alternative to getTopLevelDecls and getDeclChecked to limit which decls are deserialized by first looking at their attributes. If the attributes are accepted by a function passed as argument the decl is fully deserialized, otherwise it is ignored.

The filter is included in the signature of existing functions in the Serilalization services, but I’ve added new methods for it in FileUnit and its subclasses to leave existing implementations untouched.

This filter could be useful to address two issues:
- Included in this PR is a fix for a cycle where a general access to all top level decls lead the deserialization to call ClangImporter which triggered more deserialization. It crashed the compiler in a complex case so only limiting what is deserialized should be enough to fix this for now. rdar://problem/57118844
- Crashes in deserialization of modules using @_implementationOnly imports, most of these fail on the indexing phase trying to deserialize private decls which could be skipped if we filter by their attributes.

There are other instances of decl filtering after a call to getTopLevelDecls that are mainly filtering by Decl type. This could also be integrated in the filter functions to cover those cases and it would be worth investigating a possible performance gain.

xymus committed:
[Serialization] Filter Decl to deserialize by their attributes

Add an alternative to getTopLevelDecls and getDeclChecked to limit which
decls are deserialized by first looking at their attributes. If the
attributes are accepted by a function passed as argument the decl is
fully deserialized, otherwise it is ignored.

The filter is included in the signature of existing functions in the
Serilalization services, but I’ve added new methods for it in FileUnit
and its subclasses to leave existing implementations untouched.

xymus committed:
[ClangImporter] Deserialize only decls with a matching `@objc` name

Prevents a cycle where a general access to all top level decls lead the
deserialization to call ClangImporter which triggered more
deserialization. It crashed the compiler in a complex case so only
limiting what is deserialized should be enough to fix this for now.

rdar://problem/57118844

xymus commented:
@swift-ci Please test

xymus commented:
@swift-ci Please test source compatibility

nkcsgexi commented on include/swift/AST/FileUnit.h:
```diff
@@ -152,6 +152,20 @@ class FileUnit : public DeclContext {
+  virtual void
+  getTopLevelDeclsWhereAttributesMatch(
+              SmallVectorImpl<Decl*> &Results,
+              llvm::function_ref<bool(DeclAttributes)> matchAttributes) const;
```
Why not make it more general like `getTopLevelDeclsWhere(SmallVectorImpl<Decl*> &Results, llvm::function_ref<bool(Decl*)> match)`. We don't get much benefits by limiting it to attribute matches only.

xymus replied:
The more general function would not work in the deserialization implementation where we want to look at the attributes before deserializing the actual Decl*. Deserializing a Decl deserializes all of it, parents, members, etc. This extra work has been causing issues and we can avoid it by deserializing only the attributes to decide if it is worth a full deserialization which will give use the Decl*.

But I see the need for a more general check. I'm considering extending the matching function for more information available before the full decl: the DeclKind and maybe its name (which would require to serialize it earlier).

nkcsgexi replied:
It makes sense. Thank you for the explanation, Alexis! Could you add some doc-comments here about this reason?

xymus replied:
Sure! I've added details on how the SerializedASTFile subclass uses it. I kept it simple as to not promise too much.

nkcsgexi commented on lib/Serialization/ModuleFile.h:
```diff
@@ -888,8 +896,15 @@ class ModuleFile
-  getDeclChecked(serialization::DeclID DID);
+  getDeclChecked(
+    serialization::DeclID DID,
+    llvm::function_ref<bool(DeclAttributes)> matchAttributes = nullptr);
```
Do we need to pass down this to the function call? can we check at the caller side?

xymus replied:
To add to my answer above, the function is used in the middle of the deserialization operation, after deserializing the attributes and before the full Decl.

xymus commented:
@swift-ci Please smoke test

xymus committed:
[Serialization] Clarify role of matchAttributes in FileUnit

The subclasses to FileUnit uses the functionref matchAttributes
differently, clarify how only SerializedASTFile uses the functionref to
limit deserialization work.

xymus commented:
@swift-ci Please smoke test
